File: solutions/java/transpose/1/src/main/java/Transpose.java
import java.util.Arrays;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Transpose {

    private static final Pattern NEW_LINE_PATTERN = Pattern.compile("\\R");
    public String transpose(String input) {
        if (input.isEmpty()) {
            return "";
        }

        String[] rows = NEW_LINE_PATTERN.split(input, -1);
        int maximumWidth = this.getMaximumWidth(rows);
        return this.buildTransposedOutput(maximumWidth, rows);
    }

    private int getMaximumWidth(final String[] rows) {
        return Arrays.stream(rows)
                     .mapToInt(String::length)
                     .max()
                     .orElse(0);
    }

    private String buildTransposedOutput(final int maximumWidth, final String[] rows) {
        return IntStream.range(0, maximumWidth)
                        .mapToObj(column -> this.transposeColumn(rows, column))
                        .collect(Collectors.joining("\n"));
    }

    private String transposeColumn(String[] rows, int column) {
        int lastRow = this.findLastRowContainingColumn(rows, column);
        return IntStream.rangeClosed(0, lastRow)
                        .mapToObj(row -> this.characterAt(rows[row], column))
                        .collect(Collectors.joining());
    }

    private int findLastRowContainingColumn(String[] rows, int column) {
        return IntStream.range(0, rows.length)
                        .filter(row -> column < rows[row].length())
                        .max()
                        .orElse(-1);
    }

    private String characterAt(String row, int column) {
        return (column < row.length())
            ? String.valueOf(row.charAt(column))
            : " ";
    }
}

File: solutions/python/transpose/1/transpose.py
"""Transpose text."""

def transpose(text: str) -> str:
    """Return the transposed representation of the input text."""
    if not text:
        return ""

    rows = text.split("\n")
    maximum_length = max(len(row) for row in rows)
    transposed_rows: list[str] = []
    for column in range(maximum_length):
        transposed_rows.append(_transpose_column(rows, column))

    return "\n".join(transposed_rows)


def _transpose_column(rows: list[str], column: int) -> str:
    """Transpose one column of the input rows."""
    last_row = _find_last_row_with_character(rows, column)
    characters: list[str] = []
    for row in range(last_row + 1):
        if column < len(rows[row]):
            characters.append(rows[row][column])
        else:
            characters.append(" ")

    return "".join(characters)


def _find_last_row_with_character(rows: list[str], column: int) -> int:
    """Return the last row containing a character at the given column."""
    for row in range(len(rows) - 1, -1, -1):
        if column < len(rows[row]):
            return row

    return -1
